In [2]:
import pandas as pd
import numpy as np
from time import time
import warnings
from geopy.geocoders import Nominatim
from tqdm import tqdm
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import font_manager, rc

In [2]:
tqdm.pandas()
warnings.filterwarnings(action='ignore')
font_path = "C:/Windows/Fonts/H2GTRM.TTF"
font_name = font_manager.FontProperties(fname=font_path).get_name()
plt.rc('font',family=font_name)

In [3]:
start_time = time()
통신판매업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_08_26_04_P_통신판매업.csv",encoding = 'cp949')
일반음식점 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_24_04_P_일반음식점.csv",encoding = 'cp949')
즉석판매제조가공업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_22_19_P_즉석판매제조가공업.csv",encoding = 'cp949')
휴게음식점 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_24_05_P_휴게음식점.csv",encoding = 'cp949')
건강기능식품일반판매업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_22_03_P_건강기능식품일반판매업.csv",encoding = 'cp949')
미용업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_05_18_01_P_미용업.csv",encoding = 'cp949')
식품자동판매기업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_22_10_P_식품자동판매기업.csv",encoding = 'cp949')
담배소매업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_11_43_02_P_담배소매업.csv",encoding = 'cp949')
축산우유판매업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_22_04_P_축산판매업.csv",encoding = 'cp949')
인터넷컴퓨터게임시설제공업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_03_05_05_P_인터넷컴퓨터게임시설제공업.csv",encoding = 'cp949')
세탁업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_06_20_01_P_세탁업.csv",encoding = 'cp949')
노래연습장업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_03_09_01_P_노래연습장업.csv",encoding = 'cp949')
제과점영업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_22_18_P_제과점영업.csv",encoding = 'cp949')
단란주점영업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_23_01_P_단란주점영업.csv",encoding = 'cp949')
숙박업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_03_11_03_P_숙박업.csv",encoding = 'cp949')
체력단련장업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_10_42_01_P_체력단련장업.csv",encoding = 'cp949')
유흥주점영업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_07_23_02_P_유흥주점영업.csv",encoding = 'cp949')
골프연습장업 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_10_31_01_P_골프연습장업.csv",encoding = 'cp949')
공연장 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_03_06_01_P_공연장.csv",encoding = 'cp949')
영화상영관 = pd.read_csv("C:/Users/whileduck/Desktop/6110000_CSV/6110000_서울특별시_03_13_02_P_영화상영관.csv",encoding = 'cp949')

print(time() - start_time)

16.36482262611389


In [4]:
display(통신판매업)
print(통신판매업.columns)

번호 개방서비스명    개방서비스아이디  개방자치단체코드                   관리번호  \
0            1  통신판매업  08_26_04_P   3000000  2018300000000000000.0   
1            2  통신판매업  08_26_04_P   3000000  2018300000000000000.0   
2            3  통신판매업  08_26_04_P   3000000  2005300000000000000.0   
3            4  통신판매업  08_26_04_P   3000000  2018300000000000000.0   
4            5  통신판매업  08_26_04_P   3000000  2018300000000000000.0   
...        ...    ...         ...       ...                    ...   
633459  633460  통신판매업  08_26_04_P   3240000  2014320000000000000.0   
633460  633461  통신판매업  08_26_04_P   3240000  2013320000000000000.0   
633461  633462  통신판매업  08_26_04_P   3240000  2014320000000000000.0   
633462  633463  통신판매업  08_26_04_P   3240000  2011320000000000000.0   
633463  633464  통신판매업  08_26_04_P   3240000  2010320000000000000.0   

             인허가일자 인허가취소일자  영업상태구분코드           영업상태명 상세영업상태코드  ...  \
0       2018-11-07     NaN         1           영업/정상        1  ...   
1       2018-11-22     NaN         1           영업/정상        1  ...   
2       2005-02-17     NaN         1           영업/정상        1  ...   
3       2018-11-28     NaN         1           영업/정상        1  ...   
4       2018-11-08     NaN         1           영업/정상        1  ...   
...            ...     ...       ...             ...      ...  ...   
633459  2014-06-10     NaN         4  취소/말소/만료/정지/중지        7  ...   
633460  2011-10-06     NaN         4  취소/말소/만료/정지/중지        7  ...   
633461  2014-04-17     NaN         4  취소/말소/만료/정지/중지        7  ...   
633462  2011-08-29     NaN         4  취소/말소/만료/정지/중지        7  ...   
633463  2010-11-15     NaN         4  취소/말소/만료/정지/중지        7  ...   

                  최종수정시점 데이터갱신구분           데이터갱신일자        업태구분명      좌표정보(x)  \
0        2018-11-08 9:50       I   2018-11-10 2:37  의류/패션/잡화/뷰티  198501.5985   
1       2023-01-27 20:36       U   2023-01-29 2:40           기타  198691.7994   
2        2022-06-21 8:57       U   2022-06-23 2:40          종합몰  198691.7994   
3        2020-07-10 8:43       U   2020-07-12 2:40           기타  197567.7478   
4       2018-11-08 16:32       I   2018-11-10 2:37     레져/여행/공연   197700.034   
...                  ...     ...               ...          ...          ...   
633459  2022-11-23 17:47       U   2022-11-25 2:40  의류/패션/잡화/뷰티  211480.7664   
633460  2022-11-21 16:25       U   2022-11-23 2:40  의류/패션/잡화/뷰티  211667.3408   
633461  2022-11-22 17:04       U   2022-11-24 2:40  의류/패션/잡화/뷰티  211327.5289   
633462  2022-10-06 16:46       U   2022-10-08 2:40  의류/패션/잡화/뷰티  211719.7834   
633463  2012-03-14 15:28       I  2018-08-31 23:59        건강/식품  211459.5624   

            좌표정보(y)  자산규모 부채총액  자본금 판매방식명  
0       452850.6959   0.0  0.0  0.0   인터넷  
1       452546.5476   0.0  0.0  0.0   인터넷  
2       452546.5476   0.0  0.0  0.0   인터넷  
3       452347.9739   0.0  0.0  0.0   인터넷  
4       452022.5035   0.0  0.0  0.0   인터넷  
...             ...   ...  ...  ...   ...  
633459  450346.1630   0.0  0.0  0.0   인터넷  
633460  447647.4640   0.0  0.0  0.0   인터넷  
633461  448097.8151   0.0  0.0  0.0   인터넷  
633462  447380.2709   0.0  0.0  0.0   인터넷  
633463  447007.2852   0.0  0.0  0.0   인터넷  

[633464 rows x 32 columns]

Index(['번호', '개방서비스명', '개방서비스아이디', '개방자치단체코드', '관리번호', '인허가일자', '인허가취소일자',
       '영업상태구분코드', '영업상태명', '상세영업상태코드', '상세영업상태명', '폐업일자', '휴업시작일자', '휴업종료일자',
       '재개업일자', '소재지전화', '소재지면적', '소재지우편번호', '소재지전체주소', '도로명전체주소', '도로명우편번호',
       '사업장명', '최종수정시점', '데이터갱신구분', '데이터갱신일자', '업태구분명', '좌표정보(x)', '좌표정보(y)',
       '자산규모', '부채총액', '자본금', '판매방식명'],
      dtype='object')


In [5]:
print(통신판매업.columns)

Index(['번호', '개방서비스명', '개방서비스아이디', '개방자치단체코드', '관리번호', '인허가일자', '인허가취소일자',
       '영업상태구분코드', '영업상태명', '상세영업상태코드', '상세영업상태명', '폐업일자', '휴업시작일자', '휴업종료일자',
       '재개업일자', '소재지전화', '소재지면적', '소재지우편번호', '소재지전체주소', '도로명전체주소', '도로명우편번호',
       '사업장명', '최종수정시점', '데이터갱신구분', '데이터갱신일자', '업태구분명', '좌표정보(x)', '좌표정보(y)',
       '자산규모', '부채총액', '자본금', '판매방식명'],
      dtype='object')


# 불필요한 컬럼 제거 

In [6]:
cols = ['개방서비스명','업태구분명','사업장명','영업상태명','소재지전체주소']
cols

['개방서비스명', '업태구분명', '사업장명', '영업상태명', '소재지전체주소']

In [7]:
df_list = [통신판매업,일반음식점,즉석판매제조가공업,
건강기능식품일반판매업,
미용업,
식품자동판매기업,
담배소매업,
축산우유판매업,
인터넷컴퓨터게임시설제공업, 
노래연습장업,
제과점영업,
단란주점영업,
숙박업,
체력단련장업,
유흥주점영업,
골프연습장업,
공연장,
영화상영관]

In [8]:
df_basket = []
for df in df_list:
    new_df = df[cols]
    df_basket.append(new_df)
    
df_basket

[       개방서비스명        업태구분명                          사업장명           영업상태명  \
 0       통신판매업  의류/패션/잡화/뷰티              코지홈 (Cojee Home)           영업/정상   
 1       통신판매업           기타                   삼아씨디에스 유한회사           영업/정상   
 2       통신판매업          종합몰                      (주)코베이옥션           영업/정상   
 3       통신판매업           기타  주식회사 유키바나 (YUKIBANA Co. Ltd)           영업/정상   
 4       통신판매업     레져/여행/공연                 디드림에어서비스 주식회사           영업/정상   
 ...       ...          ...                           ...             ...   
 633459  통신판매업  의류/패션/잡화/뷰티                           바이수  취소/말소/만료/정지/중지   
 633460  통신판매업  의류/패션/잡화/뷰티               Romantic Museum  취소/말소/만료/정지/중지   
 633461  통신판매업  의류/패션/잡화/뷰티             아이엠지연 (iamjiyeon)  취소/말소/만료/정지/중지   
 633462  통신판매업  의류/패션/잡화/뷰티                         신한에스비  취소/말소/만료/정지/중지   
 633463  통신판매업        건강/식품                       허브애다이어트  취소/말소/만료/정지/중지   
 
                                   소재지전체주소  
 0                      서울특별시

In [9]:
df = pd.concat(df_basket).reset_index(drop = True)
df = df.loc[df['영업상태명'] == '영업/정상'] #현재 영업중인 데이터만 가지고 사용하도록 함 

In [10]:
df.isna().sum() #소재지를 알 수 없는 경우엔 사용하지 않음 

개방서비스명         0
업태구분명      65263
사업장명           1
영업상태명          0
소재지전체주소    45720
dtype: int64

In [11]:
na_index = df[df['소재지전체주소'].isna()].index 
df = df.drop(index = na_index,axis = 0)
df

개방서비스명        업태구분명                          사업장명  영업상태명  \
0        통신판매업  의류/패션/잡화/뷰티              코지홈 (Cojee Home)  영업/정상   
1        통신판매업           기타                   삼아씨디에스 유한회사  영업/정상   
2        통신판매업          종합몰                      (주)코베이옥션  영업/정상   
3        통신판매업           기타  주식회사 유키바나 (YUKIBANA Co. Ltd)  영업/정상   
4        통신판매업     레져/여행/공연                 디드림에어서비스 주식회사  영업/정상   
...        ...          ...                           ...    ...   
1693068  영화상영관          NaN                     CGV 천호 3관  영업/정상   
1693069  영화상영관          NaN                     CGV 천호 4관  영업/정상   
1693070  영화상영관          NaN                     CGV 천호 5관  영업/정상   
1693071  영화상영관          NaN                     CGV 천호 6관  영업/정상   
1693072  영화상영관          NaN                     CGV 천호 1관  영업/정상   

                                   소재지전체주소  
0                       서울특별시 종로구 안국동 **번지  
1                  서울특별시 종로구 경운동 **번지 수운회관  
2                    서울특별시 종로구 경운동 ** 수운회관  
3         서울특별시 종로구 내수동 **번지 용비어천가 *층 ***호  
4        서울특별시 종로구 신문로*가 **번지 고려빌딩 *층 ***호  
...                                    ...  
1693068                   서울특별시 강동구 천호동 42  
1693069                   서울특별시 강동구 천호동 42  
1693070                   서울특별시 강동구 천호동 42  
1693071                   서울특별시 강동구 천호동 42  
1693072                   서울특별시 강동구 천호동 42  

[568314 rows x 5 columns]

# 소재지를 위도, 경도로 변경해주기 위해 동까지만 사용하도록 분리 

In [12]:
def splited_addres(x):
    return ' '.join(x.split(' ')[:3])

In [13]:
df['소재지전체주소'] = df['소재지전체주소'].map(splited_addres)
df

개방서비스명        업태구분명                          사업장명  영업상태명  \
0        통신판매업  의류/패션/잡화/뷰티              코지홈 (Cojee Home)  영업/정상   
1        통신판매업           기타                   삼아씨디에스 유한회사  영업/정상   
2        통신판매업          종합몰                      (주)코베이옥션  영업/정상   
3        통신판매업           기타  주식회사 유키바나 (YUKIBANA Co. Ltd)  영업/정상   
4        통신판매업     레져/여행/공연                 디드림에어서비스 주식회사  영업/정상   
...        ...          ...                           ...    ...   
1693068  영화상영관          NaN                     CGV 천호 3관  영업/정상   
1693069  영화상영관          NaN                     CGV 천호 4관  영업/정상   
1693070  영화상영관          NaN                     CGV 천호 5관  영업/정상   
1693071  영화상영관          NaN                     CGV 천호 6관  영업/정상   
1693072  영화상영관          NaN                     CGV 천호 1관  영업/정상   

                 소재지전체주소  
0          서울특별시 종로구 안국동  
1          서울특별시 종로구 경운동  
2          서울특별시 종로구 경운동  
3          서울특별시 종로구 내수동  
4        서울특별시 종로구 신문로*가  
...                  ...  
1693068    서울특별시 강동구 천호동  
1693069    서울특별시 강동구 천호동  
1693070    서울특별시 강동구 천호동  
1693071    서울특별시 강동구 천호동  
1693072    서울특별시 강동구 천호동  

[568314 rows x 5 columns]

# 주소에서 경도, 위도 추출

### 소재지 전체주소 둘러보기

In [14]:
df['소재지전체주소'].value_counts()

서울특별시 강남구 역삼동              14459
서울특별시 서초구 서초동              14102
서울특별시 구로구 구로동              11598
서울특별시 관악구 신림동              10689
서울특별시 금천구 가산동              10677
                           ...  
서울특별시 영등포구 영등포동*가***-**        1
***-* 대화빌딩 ***호                1
서울특별시 영등포구 당산동*가***-***        1
서울특별시 영등포구 당산동*가**-***         1
경기도 이천시 대월면                    1
Name: 소재지전체주소, Length: 1814, dtype: int64

##### 총 1814 개의 유니크한 갯수들이 있고, 소수의 데이터들을 보면 서울특별시가 아니거나, 주소가 모두 가명처리 되어 있거나 되어있음 

In [15]:
not_seoul = df[~df['소재지전체주소'].str.contains('서울')] #서울 특별시가 아닌 경우도 있음 
not_seoul

개방서비스명              업태구분명          사업장명  영업상태명        소재지전체주소
8931      통신판매업        의류/패션/잡화/뷰티  카이로스(KAIROS)  영업/정상  인천광역시 계양구 작전동
27958     통신판매업                 가전         서울E&G  영업/정상   신당동 ***번지 *호
57865     통신판매업                 기타  이(e) 통인익스프레스  영업/정상  이촌동 ***-** *층
84346     통신판매업     종합몰 기타 가구/수납용품       (주)모노데코  영업/정상    경기도 김포시 대곶면
84511     통신판매업  교육/도서/완구/오락 건강/식품  주식회사 씨에프인컴퍼니  영업/정상  행당동 ***번지 **호
...         ...                ...           ...    ...            ...
1341943     미용업              일반미용업       리브가 미용실  영업/정상        130-1번지
1356710     미용업              네일아트업          쉼표네일  영업/정상   충청남도 홍성군 금마면
1602991   축산판매업              식육판매업          현대닭집  영업/정상        852- 21
1644939  노래연습장업                NaN            방방  영업/정상          0-0번지
1646688   제과점영업              제과점영업       하모니베이커리  영업/정상    경기도 이천시 대월면

[1270 rows x 5 columns]

In [16]:
df[df['소재지전체주소'].str.contains('\*')]

개방서비스명        업태구분명           사업장명  영업상태명          소재지전체주소
4              통신판매업     레져/여행/공연  디드림에어서비스 주식회사  영업/정상  서울특별시 종로구 신문로*가
5              통신판매업     컴퓨터/사무용품          제이에스티  영업/정상   서울특별시 종로구 명륜*가
10             통신판매업           기타             핑갈  영업/정상   서울특별시 종로구 종로*가
17             통신판매업  교육/도서/완구/오락   주식회사 한국동산감정원  영업/정상   서울특별시 종로구 종로*가
20             통신판매업        건강/식품        약초마을엑기스  영업/정상   서울특별시 종로구 종로*가
...              ...          ...            ...    ...              ...
1271848  건강기능식품일반판매업          NaN           미미형님  영업/정상  서울특별시 성북구 보문동*가
1271860  건강기능식품일반판매업          NaN          버터플라이  영업/정상  서울특별시 용산구 청파동*가
1271889  건강기능식품일반판매업          NaN         인셀덤용산탑  영업/정상  서울특별시 용산구 원효로*가
1271901  건강기능식품일반판매업          NaN          이현인셀덤  영업/정상  서울특별시 성북구 보문동*가
1271902  건강기능식품일반판매업          NaN            블로썸  영업/정상  서울특별시 성북구 보문동*가

[34300 rows x 5 columns]

In [3]:
def get_address(address):
    
    '''
    geopy 모듈을 이용하여 도로명 주소에서 경도, 위도를 담은
    자료로 변경해주는 함수 
    '''
    
    

    geolocoder = Nominatim(user_agent = 'South Korea', timeout = None)
    geo = geolocoder.geocode(address)
    
    return geo

In [18]:
print(get_address('종로2가사거리정류소'))

None


In [19]:
def get_lat(x):
    return x.latitude
def get_lot(x):
    return x.longitude

In [20]:
print(get_address('서울특별시 성북구 보문동*가'))
print(get_address('서울특별시 성북구 명륜*가'))

# 보게 되면 *이 포함 되어 있어도 조사가 되는 경우가 있고, 되지 않는 경우가 있다. 
# 우선 이럴 때는 전부 위, 경도를 구한 후에 타입이 있는 경우에 변경해주자 

보문동1가, 보문동, 성북구, 서울, 02870, 대한민국
None


In [21]:
df = df.drop(index = not_seoul.index , axis = 0) #서울이 아닌 곳 제거 
df

개방서비스명        업태구분명                          사업장명  영업상태명  \
0        통신판매업  의류/패션/잡화/뷰티              코지홈 (Cojee Home)  영업/정상   
1        통신판매업           기타                   삼아씨디에스 유한회사  영업/정상   
2        통신판매업          종합몰                      (주)코베이옥션  영업/정상   
3        통신판매업           기타  주식회사 유키바나 (YUKIBANA Co. Ltd)  영업/정상   
4        통신판매업     레져/여행/공연                 디드림에어서비스 주식회사  영업/정상   
...        ...          ...                           ...    ...   
1693068  영화상영관          NaN                     CGV 천호 3관  영업/정상   
1693069  영화상영관          NaN                     CGV 천호 4관  영업/정상   
1693070  영화상영관          NaN                     CGV 천호 5관  영업/정상   
1693071  영화상영관          NaN                     CGV 천호 6관  영업/정상   
1693072  영화상영관          NaN                     CGV 천호 1관  영업/정상   

                 소재지전체주소  
0          서울특별시 종로구 안국동  
1          서울특별시 종로구 경운동  
2          서울특별시 종로구 경운동  
3          서울특별시 종로구 내수동  
4        서울특별시 종로구 신문로*가  
...                  ...  
1693068    서울특별시 강동구 천호동  
1693069    서울특별시 강동구 천호동  
1693070    서울특별시 강동구 천호동  
1693071    서울특별시 강동구 천호동  
1693072    서울특별시 강동구 천호동  

[567044 rows x 5 columns]

In [22]:
def get_address(address):
    
    '''
    geopy 모듈을 이용하여 도로명 주소에서 경도, 위도를 담은
    자료로 변경해주는 함수 
    '''
    
    

    geolocoder = Nominatim(user_agent = 'South Korea', timeout = None)
    geo = geolocoder.geocode(address)
    
    return geo

# 53만개의 데이터를 모두 적용하기 보다, 각 유니크한 값의 경도, 위도를 추출 한 후 
# map 을 이용하여 변경해주도록 함 

In [23]:
location = pd.Series(df['소재지전체주소'].unique())
address = location.progress_map(get_address)
address

100%|██████████████████████████████████████████████████████████████████████████████| 1394/1394 [23:18<00:00,  1.00s/it]


0       (안국동, 삼청동, 종로구, 서울, 03161, 대한민국, (37.57721, 12...
1       (경운동, 종로1·2·3·4가동, 종로구, 서울, 03131, 대한민국, (37.5...
2       (내수동, 사직동, 종로구, 서울, 03173, 대한민국, (37.57294, 12...
3                                                    None
4                                                    None
                              ...                        
1389    (암사2동, 강동구, 서울, 05242, 대한민국, (37.55816585, 127...
1390                   (강동구, 서울, 대한민국, (37.53, 127.1237))
1391    (고덕2동, 강동구, 서울, 대한민국, (37.5658618, 127.1638516...
1392    (천호옛길, 성내2동, 강동구, 서울, 05391, 대한민국, (37.5341, 1...
1393    (모진동, 화양동, 광진구, 서울, 05029, 대한민국, (37.54185, 12...
Length: 1394, dtype: object

In [24]:
none_type_num = address[address.isna()].index #주소가 nonetype 인 경우의 인덱스 번호 저장 
none_type_value = location[none_type_num] #주소가 nonetype 인 경우의 주소 저장 
new_location = location.drop(index = none_type_num, axis = 0)

new_location

0        서울특별시 종로구 안국동
1        서울특별시 종로구 경운동
2        서울특별시 종로구 내수동
5        서울특별시 종로구 봉익동
6        서울특별시 종로구 창신동
             ...      
1389    서울특별시 강동구 암사2동
1390        서울특별시 강동구 
1391    서울특별시 강동구 고덕2동
1392    서울특별시 강동구 천호옛길
1393     서울특별시 광진구 모진동
Length: 1269, dtype: object

In [25]:
none_type_index = df.loc[df['소재지전체주소'].isin(none_type_value)].index #주소가 nonetype인 데이터 프레임의 인덱스
df = df.drop(index = none_type_index, axis = 0) #주소가 nonetype인 데이터 프레임의 인덱스 제거 
df

개방서비스명        업태구분명                          사업장명  영업상태명  \
0        통신판매업  의류/패션/잡화/뷰티              코지홈 (Cojee Home)  영업/정상   
1        통신판매업           기타                   삼아씨디에스 유한회사  영업/정상   
2        통신판매업          종합몰                      (주)코베이옥션  영업/정상   
3        통신판매업           기타  주식회사 유키바나 (YUKIBANA Co. Ltd)  영업/정상   
6        통신판매업  의류/패션/잡화/뷰티                        조은 주얼리  영업/정상   
...        ...          ...                           ...    ...   
1693068  영화상영관          NaN                     CGV 천호 3관  영업/정상   
1693069  영화상영관          NaN                     CGV 천호 4관  영업/정상   
1693070  영화상영관          NaN                     CGV 천호 5관  영업/정상   
1693071  영화상영관          NaN                     CGV 천호 6관  영업/정상   
1693072  영화상영관          NaN                     CGV 천호 1관  영업/정상   

               소재지전체주소  
0        서울특별시 종로구 안국동  
1        서울특별시 종로구 경운동  
2        서울특별시 종로구 경운동  
3        서울특별시 종로구 내수동  
6        서울특별시 종로구 봉익동  
...                ...  
1693068  서울특별시 강동구 천호동  
1693069  서울특별시 강동구 천호동  
1693070  서울특별시 강동구 천호동  
1693071  서울특별시 강동구 천호동  
1693072  서울특별시 강동구 천호동  

[555467 rows x 5 columns]

In [26]:
new_address = address.drop(index = none_type_num)
new_address

0       (안국동, 삼청동, 종로구, 서울, 03161, 대한민국, (37.57721, 12...
1       (경운동, 종로1·2·3·4가동, 종로구, 서울, 03131, 대한민국, (37.5...
2       (내수동, 사직동, 종로구, 서울, 03173, 대한민국, (37.57294, 12...
5       (봉익동, 종로1·2·3·4가동, 종로구, 서울, 03139, 대한민국, (37.5...
6       (창신동, 창신3동, 종로구, 서울, 03103, 대한민국, (37.57595, 1...
                              ...                        
1389    (암사2동, 강동구, 서울, 05242, 대한민국, (37.55816585, 127...
1390                   (강동구, 서울, 대한민국, (37.53, 127.1237))
1391    (고덕2동, 강동구, 서울, 대한민국, (37.5658618, 127.1638516...
1392    (천호옛길, 성내2동, 강동구, 서울, 05391, 대한민국, (37.5341, 1...
1393    (모진동, 화양동, 광진구, 서울, 05029, 대한민국, (37.54185, 12...
Length: 1269, dtype: object

In [27]:
lat = new_address.progress_map(get_lat)
lot = new_address.progress_map(get_lot)

100%|██████████████████████████████████████████████████████████████████████████| 1269/1269 [00:00<00:00, 568043.95it/s]


In [28]:
def make_dict(x,y):
    
    X = np.array(x).reshape(-1,1)
    Y = np.array(y).reshape(-1,1)
    
    return dict(np.c_[X,Y])

In [29]:
dicted_lat = make_dict(new_location,lat)
dicted_lot = make_dict(new_location,lot)
dicted_lat

{'서울특별시 종로구 안국동': 37.57721,
 '서울특별시 종로구 경운동': 37.57489,
 '서울특별시 종로구 내수동': 37.57294,
 '서울특별시 종로구 봉익동': 37.572,
 '서울특별시 종로구 창신동': 37.57595,
 '서울특별시 종로구 팔판동': 37.5825315,
 '서울특별시 종로구 평창동': 37.611446,
 '서울특별시 종로구 종로*가': 37.5704405,
 '서울특별시 종로구 관수동': 37.5688,
 '서울특별시 종로구 통인동': 37.5800949,
 '서울특별시 종로구 묘동': 37.5723634,
 '서울특별시 종로구 부암동': 37.5949602,
 '서울특별시 종로구 숭인동': 37.57648,
 '서울특별시 종로구 계동': 37.5811587,
 '서울특별시 종로구 효제동': 37.5732,
 '서울특별시 종로구 구기동': 37.609029,
 '서울특별시 종로구 관철동': 37.56878,
 '서울특별시 종로구 원서동': 37.58146,
 '서울특별시 종로구 관훈동': 37.5742524,
 '서울특별시 종로구 인의동': 37.57236,
 '서울특별시 종로구 원남동': 37.57538,
 '서울특별시 종로구 권농동': 37.57537,
 '서울특별시 종로구 장사동': 37.56897,
 '서울특별시 종로구 서린동': 37.56925,
 '서울특별시 종로구 이화동': 37.57663,
 '서울특별시 종로구 연건동': 37.5784056,
 '서울특별시 종로구 동숭동': 37.58096,
 '서울특별시 종로구 혜화동': 37.58755,
 '서울특별시 종로구 견지동': 37.57435,
 '서울특별시 종로구 옥인동': 37.58348,
 '서울특별시 종로구 청운동': 37.58895,
 '서울특별시 종로구 평동': 37.5683558,
 '서울특별시 종로구 화동': 37.580674200000004,
 '서울특별시 종로구 예지동': 37.5695591,
 '서울특별시 종로구 익선동': 37.57

In [30]:
df['lat'] = df['소재지전체주소'].map(dicted_lat)
df['lot'] = df['소재지전체주소'].map(dicted_lot)
df

개방서비스명        업태구분명                          사업장명  영업상태명  \
0        통신판매업  의류/패션/잡화/뷰티              코지홈 (Cojee Home)  영업/정상   
1        통신판매업           기타                   삼아씨디에스 유한회사  영업/정상   
2        통신판매업          종합몰                      (주)코베이옥션  영업/정상   
3        통신판매업           기타  주식회사 유키바나 (YUKIBANA Co. Ltd)  영업/정상   
6        통신판매업  의류/패션/잡화/뷰티                        조은 주얼리  영업/정상   
...        ...          ...                           ...    ...   
1693068  영화상영관          NaN                     CGV 천호 3관  영업/정상   
1693069  영화상영관          NaN                     CGV 천호 4관  영업/정상   
1693070  영화상영관          NaN                     CGV 천호 5관  영업/정상   
1693071  영화상영관          NaN                     CGV 천호 6관  영업/정상   
1693072  영화상영관          NaN                     CGV 천호 1관  영업/정상   

               소재지전체주소       lat        lot  
0        서울특별시 종로구 안국동  37.57721  126.98363  
1        서울특별시 종로구 경운동  37.57489  126.98625  
2        서울특별시 종로구 경운동  37.57489  126.98625  
3        서울특별시 종로구 내수동  37.57294  126.97180  
6        서울특별시 종로구 봉익동  37.57200  126.99303  
...                ...       ...        ...  
1693068  서울특별시 강동구 천호동  37.54294  127.12733  
1693069  서울특별시 강동구 천호동  37.54294  127.12733  
1693070  서울특별시 강동구 천호동  37.54294  127.12733  
1693071  서울특별시 강동구 천호동  37.54294  127.12733  
1693072  서울특별시 강동구 천호동  37.54294  127.12733  

[555467 rows x 7 columns]

# 결측치 처리 

In [31]:
df.isna().sum()

개방서비스명         0
업태구분명      63436
사업장명           1
영업상태명          0
소재지전체주소        0
lat            0
lot            0
dtype: int64

In [32]:
df.loc[df['업태구분명'].isna(),'개방서비스명'].value_counts()

건강기능식품일반판매업      30199
담배소매업            18178
노래연습장업            5606
체력단련장업            3330
인터넷컴퓨터게임시설제공업     3015
골프연습장업            1598
공연장                841
영화상영관              648
즉석판매제조가공업           12
일반음식점                9
Name: 개방서비스명, dtype: int64

### 확인 결과 일반음식점 제조업을 제외한 다른 것들은, 업태구분명이 모두 없는 것으로 판명되었다.
### 일반음식점은 데이터를 둘러보고 결측치를 변경해주고, 그 이상의 것들은 데이터의 특성에 따라 변경해주자 

In [33]:
cols = df.loc[df['업태구분명'].isna(),'개방서비스명'].value_counts().index.tolist()[:-1]

for _ in cols:
    cond = (df['업태구분명'].isna()) & (df['개방서비스명'] == _)
    df.loc[cond,'업태구분명'] = df.loc[cond,'업태구분명'].fillna(df.loc[cond,'개방서비스명']) #개방서비스명으로 업태구분명을 채워줌

In [34]:
df.loc[(df['업태구분명'].isna())& (df['개방서비스명'] == '일반음식점')] #각자 해당하는 업태구분명이 달라보인다.

개방서비스명 업태구분명              사업장명  영업상태명        소재지전체주소        lat  \
637051  일반음식점   NaN           대독장 교대점  영업/정상  서울특별시 서초구 서초동  37.488441   
643860  일반음식점   NaN           을지로 골뱅이  영업/정상  서울특별시 종로구 관수동  37.568800   
654100  일반음식점   NaN         콩심 신정네거리점  영업/정상  서울특별시 양천구 신정동  37.516241   
657644  일반음식점   NaN               일빵빵  영업/정상  서울특별시 종로구 동숭동  37.580960   
678606  일반음식점   NaN       고희케이터링 협동조합  영업/정상  서울특별시 종로구 창성동  37.580883   
685171  일반음식점   NaN              감성포차  영업/정상  서울특별시 강동구 천호동  37.542940   
690920  일반음식점   NaN  불닭발땡초동대문엽기떡볶이대치점  영업/정상  서울특별시 강남구 대치동  37.498920   
714175  일반음식점   NaN             인디언소울  영업/정상  서울특별시 도봉구 도봉동  37.682606   
717320  일반음식점   NaN              올까말까  영업/정상  서울특별시 종로구 홍지동  37.600410   

               lot  
637051  127.016442  
643860  126.990180  
654100  126.852120  
657644  127.005193  
678606  126.972545  
685171  127.127330  
690920  127.065786  
714175  127.042541  
717320  126.956480

In [35]:
df.loc[df['개방서비스명'] == '일반음식점','업태구분명'].unique()

array(['한식', '일식', '기타', '출장조리', '분식', '식육(숯불구이)', '중국식', '경양식', '호프/통닭',
       '까페', '정종/대포집/소주방', '뷔페식', '김밥(도시락)', '통닭(치킨)', '횟집', '패밀리레스트랑',
       '감성주점', '외국음식전문점(인도,태국등)', '라이브카페', '복어취급', '키즈카페', '냉면집', '패스트푸드',
       '탕류(보신용)', '커피숍', nan, '전통찻집', '이동조리'], dtype=object)

In [36]:
df.loc[(df['업태구분명'].isna())& (df['개방서비스명'] == '일반음식점'),'사업장명'].unique()

array(['대독장 교대점', '을지로 골뱅이', '콩심 신정네거리점', '일빵빵', '고희케이터링 협동조합', '감성포차',
       '불닭발땡초동대문엽기떡볶이대치점', '인디언소울', '올까말까'], dtype=object)

In [37]:
df.loc[(df['업태구분명'].isna())& (df['개방서비스명'] == '일반음식점'),'업태구분명']= ['한식','호프/통닭','한식','한식','한식','감성주점','분식','한식','한식']

In [38]:
df.isna().sum()

개방서비스명     0
업태구분명      0
사업장명       1
영업상태명      0
소재지전체주소    0
lat        0
lot        0
dtype: int64

# 사업장명을 모르는 경우에는 제거하고 사용하자 

In [39]:
df = df.drop(index = df[df['사업장명'].isna()].index,axis = 0)
df = df.drop('영업상태명',axis = 1).reset_index(drop = True)

In [40]:
df['구']= [x[1] for x in df['소재지전체주소'].str.split(' ')]
df

개방서비스명        업태구분명                          사업장명        소재지전체주소  \
0       통신판매업  의류/패션/잡화/뷰티              코지홈 (Cojee Home)  서울특별시 종로구 안국동   
1       통신판매업           기타                   삼아씨디에스 유한회사  서울특별시 종로구 경운동   
2       통신판매업          종합몰                      (주)코베이옥션  서울특별시 종로구 경운동   
3       통신판매업           기타  주식회사 유키바나 (YUKIBANA Co. Ltd)  서울특별시 종로구 내수동   
4       통신판매업  의류/패션/잡화/뷰티                        조은 주얼리  서울특별시 종로구 봉익동   
...       ...          ...                           ...            ...   
555461  영화상영관        영화상영관                     CGV 천호 3관  서울특별시 강동구 천호동   
555462  영화상영관        영화상영관                     CGV 천호 4관  서울특별시 강동구 천호동   
555463  영화상영관        영화상영관                     CGV 천호 5관  서울특별시 강동구 천호동   
555464  영화상영관        영화상영관                     CGV 천호 6관  서울특별시 강동구 천호동   
555465  영화상영관        영화상영관                     CGV 천호 1관  서울특별시 강동구 천호동   

             lat        lot    구  
0       37.57721  126.98363  종로구  
1       37.57489  126.98625  종로구  
2       37.57489  126.98625  종로구  
3       37.57294  126.97180  종로구  
4       37.57200  126.99303  종로구  
...          ...        ...  ...  
555461  37.54294  127.12733  강동구  
555462  37.54294  127.12733  강동구  
555463  37.54294  127.12733  강동구  
555464  37.54294  127.12733  강동구  
555465  37.54294  127.12733  강동구  

[555466 rows x 7 columns]

In [41]:
sns.set_style('whitegrid')

In [42]:
df.to_csv('서울특별시_상권.csv',index = False)

In [43]:
df = pd.read_csv('서울특별시_상권.csv')
df

개방서비스명        업태구분명                          사업장명        소재지전체주소  \
0       통신판매업  의류/패션/잡화/뷰티              코지홈 (Cojee Home)  서울특별시 종로구 안국동   
1       통신판매업           기타                   삼아씨디에스 유한회사  서울특별시 종로구 경운동   
2       통신판매업          종합몰                      (주)코베이옥션  서울특별시 종로구 경운동   
3       통신판매업           기타  주식회사 유키바나 (YUKIBANA Co. Ltd)  서울특별시 종로구 내수동   
4       통신판매업  의류/패션/잡화/뷰티                        조은 주얼리  서울특별시 종로구 봉익동   
...       ...          ...                           ...            ...   
555461  영화상영관        영화상영관                     CGV 천호 3관  서울특별시 강동구 천호동   
555462  영화상영관        영화상영관                     CGV 천호 4관  서울특별시 강동구 천호동   
555463  영화상영관        영화상영관                     CGV 천호 5관  서울특별시 강동구 천호동   
555464  영화상영관        영화상영관                     CGV 천호 6관  서울특별시 강동구 천호동   
555465  영화상영관        영화상영관                     CGV 천호 1관  서울특별시 강동구 천호동   

             lat        lot    구  
0       37.57721  126.98363  종로구  
1       37.57489  126.98625  종로구  
2       37.57489  126.98625  종로구  
3       37.57294  126.97180  종로구  
4       37.57200  126.99303  종로구  
...          ...        ...  ...  
555461  37.54294  127.12733  강동구  
555462  37.54294  127.12733  강동구  
555463  37.54294  127.12733  강동구  
555464  37.54294  127.12733  강동구  
555465  37.54294  127.12733  강동구  

[555466 rows x 7 columns]